Review

StarostinLF opened the pull request.

StarostinLF committed:
fix(.eslintrc.js): линтер теперь не ругается на конец каждой строки

StarostinLF committed:
feat: свёрстана форма и подключено состояние к ней

StarostinLF committed:
fix(articleparamsform.module.scss): добавлено расстояние в форме и убран скролл

Добавлено расстояние между полями ввода и заголовка в форме, а также убран лишний скролл формы

StarostinLF committed:
fix(articleparamsform.tsx): исправлено закрытие формы при изменении полей ввода

StarostinLF committed:
fix(src/index.tsx): исправлено отображение изменений текста

Исправлено отображение изменений текстов при применении стилей в форме

gennady-bars commented on src/index.tsx:
```diff
@@ -13,19 +13,24 @@ const domNode = document.getElementById('root') as HTMLDivElement;
+	const [articleState, setArticleState] = useState(defaultArticleState);
+
 	return (
 		<div
```
* По семантике нужно помещать весь главный контент страницы **после** хэдера в тег `main`.

gennady-bars commented on src/index.tsx:
```diff
@@ -13,19 +13,24 @@ const domNode = document.getElementById('root') as HTMLDivElement;
+
 	return (
 		<div
 			className={clsx(styles.main)}
```
**Можно лучше**

* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`
* `clsx` можно не использовать, если только **один** селектор передается

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,146 @@
+	articleState,
+	setArticleState,
+}: ArticleParamsFormProps) => {
+	const [isOpen, setIsOpen] = useState<boolean>(false);
```
* `[isOpen, setIsOpen]` нужно назвать `[isMenuOpen, setIsMenuOpen]`, чтобы было понятно, что тут открываете

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,146 @@
+	});
+	const rootRef = useRef<HTMLElement | null>(null);
+
+	useOutsideClickClose({
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для правильного навешивания и удаления обработчиков `Escape` и оверлея (клика снаружи блока):

```js
import { useEffect } from "react";

type TUseClose = {
  isOpen: boolean;
  onClose: () => void;
  rootRef: React.RefObject<HTMLElement>;
};
// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export function useClose({isOpen, onClose, rootRef}: TUseClose) {
  useEffect(() => {
    if (!isOpen) return; // останавливаем действие эффекта, если закрыто

   
    function handleClickOutside(event: MouseEvent) {
      const { target } = event;
      const isOutsideClick =
        target instanceof Node && // проверяем, что это `DOM`-элемент
        rootRef.current &&
        !rootRef.current.contains(target);  // проверяем, что кликнули на элемент, который находится не внутри нашего блока
      if (isOutsideClick) {
        onClose();
      }
    }

    const handleEscape = (e: KeyboardEvent) => {
      if (e.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    document.addEventListener("mousedown", handleClickOutside);

    //  обязательно удаляем обработчики в `clean-up`- функции
    return () => {
      document.removeEventListener("keydown", handleEscape);
      document.removeEventListener("mousedown", handleClickOutside);
    };
    // обязательно следим за `isOpen`, чтобы срабатывало только при открытии, а не при любой перерисовке компонента
  }, [isOpen, onClose, rootRef]);
}


```

Этот код помещают в отдельный файл `useClose.ts` в папке `hooks` и импортируют хук туда, где нужно.

И теперь можно использовать в любом компоненте (попапе, меню и тд), которому нужно установить эти обработчики. 

Пример того, как это использовать:
 
```js
    // создаем `ref` для нашей формы и указываем его на внешнем блоке с формой
    const formRef = useRef<HTMLElement>(null);
    
    useClose({
        isOpen: isMenuOpen,
        onClose: closeMenu,
        rootRef: formRef,
    });

```

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,146 @@
+				ref={rootRef}>
+				<form
+					className={styles.form}
+					onSubmit={formSubmitHandler}
```
Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. Обработка события `submit` еще срабатывает при нажатии на `Enter`, так как это встроенная логика сабмитов в браузере.

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,146 @@
-				<form className={styles.form}>
+			<ArrowButton onClick={setIsOpen} isOpen={isOpen} />
+			<aside
+				className={`${styles.container} ${isOpen ? styles.container_open : ''}`}
```
**Можно лучше**


* Библиотека `clsx` имеет возможность добавления стилей по **условию**: если оно будет `true`, то добавится селектор, а если `false`, то нет. Подробнее тут  https://habr.com/ru/articles/649381/

gennady-bars commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,14 +1,146 @@
+	setArticleState,
+}: ArticleParamsFormProps) => {
+	const [isOpen, setIsOpen] = useState<boolean>(false);
+	const [syleConfig, setSyleConfig] = useState({
```
`[syleConfig, setSyleConfig]` нужно назвать `[formState, setFormState]`, так как тут состояние изменений **формы**

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:

* Хорошо реализовали форму из предоставленных компонентов
* Хорошая логика передачи данных от `App` в форму и обратно
* Отлично, что обработчик сабмита вешаете на форму, а не клик на кнопку. 
* Отлично, что нет типов `any`
* Отлично типизировали проект


но есть некоторые недочеты:

* По семантике нужно помещать весь главный контент страницы **после** хэдера в тег `main`.
* `[isOpen, setOpen]` нужно назвать `[isMenuOpen, setIsMenuOpen]`, чтобы было понятно, что тут открываете
* Скрин https://disk.yandex.ru/i/X988fGbgOuy5jw стрелочка в открытом меню должна менять направление по макету 
* `[syleConfig, setSyleConfig]` нужно назвать `[formState, setFormState]`, так как тут состояние изменений **формы** 




**Можно лучше**

* Лучше выделить компонент  `App` в отдельный файл, чтобы не смешивать все в `index.js`
* `clsx` можно не использовать, если только **один** селектор передается
* Библиотека `clsx` имеет возможность добавления стилей по **условию**: если оно будет `true`, то добавится селектор, а если `false`, то нет. Подробнее тут  https://habr.com/ru/articles/649381/

StarostinLF committed:
fix(package.json): убран yarn

StarostinLF committed:
refactor(src/index.tsx): компонент App перенесён в отдельный файл

StarostinLF committed:
refactor: переименование констант хука useState

Переименованы константы хука useState isOpen и styleConfig в isMenuOpen и formState соответственно

StarostinLF committed:
refactor(articleparamsform.tsx): добавлен clsx для тега aside

StarostinLF committed:
feat(useclose.ts): добавлен кастомный хук useClose

Добавлен кастомный хук useClose для правильного навешивания и удаления обработчиков Escape и оверлея
(клика снаружи блока)

StarostinLF committed:
fix(arrowbutton.tsx): исправлено положение стрелки при открытии и закрытии формы

StarostinLF commented:
Подскажите, как можно сделать gh-pages в данном проекте? Сборка есть, но в конечном итоге на сайте приложение не отображается
![image](https://github.com/StarostinLF/blog-customizer/assets/102907077/af9515bc-67b8-4c7c-9bde-74049a4932b2)

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

gennady-bars commented:
> Подскажите, как можно сделать gh-pages в данном проекте? Сборка есть, но в конечном итоге на сайте приложение не отображается

бывает такое из-за неправильного `homepage` в файле `package.json`. Нужно это перепроверить

StarostinLF committed:
fix(package.json): исправлен homepage
